### DISCUSSION TITLE
Save bucket state in an SQL database

### DISCUSSION THREAD

**Discussion by radiKal07** [vladimir-bukhtoyarov/bucket4j#146] [Ideas]
I think it would be nice to support saving bucket state in an SQL database (or maybe an integration directly with Hibernate) for whom the in-memory solution is not sufficient.
What do you think?

**Comment by vladimir-bukhtoyarov** [comment]
@tudor07 hello,

Before opening my thoughts about RDBMs integration I want to describe current existed Bucket4j integrations. All of them are implemented with the following performance rule: If you want to achieve low latency then it is better to deliver computation to the place where data is stored instead of reading data to the place where computation happens. Maybe [this article](https://hazelcast.com/blog/an-easy-performance-improvement-with-entryprocessor/) better describes my idea. In contexts of RDBMs it is enough to use stored procedures in order to satisfy this rule. I'he created the plans for each major database like #37 #38  

But... Each RDBMs has own syntax for stored procedures, so we can not just write something once and reuse it many times. Due to this work requires significant efforts I've postponed RDBM's integrations a bit. I am going to return to this task after providing integration for Redis.

But... If you do not need optimal performance(looks like your case) you shouldn't wait for direct integration due to I've created a prototyping framework for such cases. You can play with it using the [following dependency](https://mvnrepository.com/artifact/com.github.vladimir-bukhtoyarov/bucket4j-core/5.0.0-beta.01-09-2020) I propose two options for support generic database:
* Extending [AbstractLockBasedBackend](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/generic/select_for_update/AbstractLockBasedBackend.java) when your database supports the locking for transaction isolation. Example can be found there [AdvisoryLockBasedPostgreSQLBackend](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-examples/src/test/java/example/distributed/generic/postgresql/advisory_lock/AdvisoryLockBasedPostgreSQLBackend.java)
* Extending [AbstractCompareAndSwapBasedBackend](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/generic/compare_and_swap/AbstractCompareAndSwapBasedBackend.java) when your database does not provide the locking but provides any variant of Compare&Swap idiom. An example can be found there [RMapBasedRedissonBackend](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-examples/src/test/java/example/distributed/generic/redisson/rmap_cas/RMapBasedRedissonBackend.java)

I hope that this information will be enough for you to build your own backend on top of the technology that you prefer. According to version ```0.0-beta.01-09-2020``` - it is a well-tested version, but I've not released it yet because I had not enough time last year to write documentation for this version, so be prepared for the partial absence of javadocs and readme pages
